main/java/us/klette/constantstring/CString.java

```diff
@@ -0,0 +1,143 @@
+ * copyright notice, this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided
+ * with the distribution. 3) Neither the name of the Qulice.com nor
+ * the names of its contributors may be used to endorse or promote
+ * products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT
+ * NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package us.klette.constantstring;
+
+import javax.annotation.Nonnull;
+
+/**
+ * The CString interface defines the allowed operations
+ * within ConstantString.
+ * <p>
+ * This is a small utility library for doing constant time string manipulation.
+ * Or in simpler terms, you can do operations on strings that are done in a
+ * fixed time, and delay the linear time computation until you evaluate your
+ * set of operations.
+ *
+ * @author Kristian Klette (klette@klette.us)
+ */
+public interface CString {
+
+    /**
+     * Creates a new CString instance from a String.
+     *
+     * @param value The String to use as a base value
+     * @return The CString instance from the string
+     */
+    static CString create(final String value) {
+        return new CStringLeaf(value);
+    }
+
+    /**
+     * Evaluates the CString into a String.
+     *
+     * @return The resulting {@link String} instance.
+     */
+    @Override
+    String toString();
+
+    /**
+     * Applies a substring operations similar to
+     * {@link java.lang.String#substring(int)}.
+     *
+     * @param index The index of which to start the new string
+     * @return The CString representation after the applied substring operation
+     */
+    default CString substring(final int index) {
+        return new CStringSub(this, index);
+    }
+    /**
+     * Applies a substring operations similar to
+     * {@link java.lang.String#substring(int, int)}.
+     *
+     * @param index The index of which to start the new string
+     * @param end The end index of the substring
+     * @return The CString representation after the applied substring operation
+     */
+    default CString substring(final int index, final int end) {
+        return new CStringSubRange(this, index, end);
```

Comment:
Not sure if we need both `CStringSub` and `CStringSubRange,` isn't it the same case, `substring(3)` being simply `substring(3, this.length)`? (plus minus 1 :))